### ISSUE TITLE
Lookup resource from AWS

### ISSUE THREAD

**Issue by blade2005** [cloud-custodian/cloud-custodian#5056]
I'm dealing with CloudWatch alarms and would like to find the ec2 instance name, I have the dimension in the cloudwatch with the instance id but no instance name. Is this possible? Should I submit a PR with a generic helper function to look up arbritrary information from aws?

**Comment by kapilt** [comment]
could you clarify if you mean in a policy on cloud watch alarms, or if you mean in the mailer at notification?  typically for policies, we would support a related resource filter that would do a one hop away association to the policy resource. ie. security group or image filter for an ec2 instance. cloud watch alarms are a bit interesting in that they can really be associated to any resource on the other side.

**Comment by blade2005** [comment]
in the mailer notification is what I'm struggling with. In much the same vein as my other issues, I'm notifying for active disk alarms and I'd much rather have the instance names (Instance Name == AD name) than the instance id.

**Comment by kapilt** [comment]
so the right way is probably doing this via filter that annotates the resource so that it can be available in either the policy or in the messaging template. trying to do resolution in the mailer gets a bit more convoluted on iam policies and role assumes back in, the mailer has a fairly limited ability to do that today, but its only for the purpose of delivering to an sns topic in a different account.

**Comment by blade2005** [comment]
That looks like a much heavier lift. I'm not sure how we would determine the resource type in the filter annotation code in order to get related resources correctly.

**Comment by blade2005** [comment]
To make sure I'm understanding this correctly.
 Following the example in https://github.com/cloud-custodian/cloud-custodian/blob/master/c7n/filters/vpc.py#L31-L40 we would add `RelatedResource` and `AnnotationKey` to https://github.com/cloud-custodian/cloud-custodian/blob/master/c7n/resources/cw.py#L31-L44 as well as importing the `RelatedResourceFilter` and having the class inherit from it.

Then we'd need something that gets executed based on the alarms returned and determines via :magic: what the resource type is.

**Comment by kapilt** [comment]
agreed its a heavier lift, as it would need dimension grouping association to resources, some of that is clear via id matching, but its also reasonable to punt and ask users to be explicit on the resource in the policy for the related.
